### PR TITLE
feat: parse `try-with`, `do`, `resume` expressions

### PR DESCRIPTION
**Pull request by mlutze**
related to #3542, depends on #3547

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 1]
```diff
@@ -1630,6 +1687,15 @@ object ParsedAst {
...
   case class CatchRule(ident: Name.Ident, fqn: Seq[String], exp: ParsedAst.Expression)
 
+  /**
+    * Effect handler rule.
+    *
+    * @param name    the operation name.
+    * @param fparams the operation parameters.
+    * @param exp     the body expression.
+    */
+  case class WithRule(name: Name.QName, fparams: Seq[FormalParam], exp: ParsedAst.Expression)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1697]: Lets call it `HandlerRule`.

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 2]
```diff
@@ -881,14 +917,35 @@ object ParsedAst {
...
+      * @param catchOrWith the handler (catch/with) of the try expression.
+      * @param sp2         the position of the last character in the expression.
       */
-    case class TryCatch(sp1: SourcePosition, exp: ParsedAst.Expression, rules: Seq[ParsedAst.CatchRule], sp2: SourcePosition) extends ParsedAst.Expression
+    case class Try(sp1: SourcePosition, exp: ParsedAst.Expression, catchOrWith: CatchOrWith, sp2: SourcePosition) extends ParsedAst.Expression
+
+    /**
+      * An enum representing the handler of a `try` expression.
+      */
+    sealed trait CatchOrWith
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:932]: -> CatchOrHandler

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 3]
```diff
@@ -437,6 +466,13 @@ object ParsedAst {
       */
     case class Default(sp1: SourcePosition, sp2: SourcePosition) extends ParsedAst.Literal
 
+    /**
+      * Resume Literal.
+      *
+      * @param sp1 the position of the first character in the literal.
+      * @param sp2 the position of the last character in the literal.
+      */
+    case class Resume(sp1: SourcePosition, sp2: SourcePosition) extends ParsedAst.Literal
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:475]: It should be an expression.

I think we should keep them together: Do/Try/Resume or maybe Do/Resume/Try?

Not sure where exactly in the ast they fit. They are control operators.

**Comment by magnus-madsen** [comment]
Will you also add test cases? :)

**Comment by mlutze** [comment]
Per discussion: need to change `case` syntax to `def` syntax.

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 4]
```diff
@@ -910,14 +909,54 @@ object ParsedAst {
...
+      * @param exp            the guarded expression.
+      * @param catchOrHandler the handler (catch/with) of the try expression.
+      * @param sp2            the position of the last character in the expression.
+      */
+    case class Try(sp1: SourcePosition, exp: ParsedAst.Expression, catchOrHandler: CatchOrHandler, sp2: SourcePosition) extends ParsedAst.Expression
+
+    /**
+      * An enum representing the handler of a `try` expression.
+      */
+    sealed trait CatchOrHandler
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:943]: Usually this helper types are plased later in the file right? Not in here in Expression?

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 5]
```diff
@@ -910,14 +909,54 @@ object ParsedAst {
...
+
+    object CatchOrHandler {
+      /**
+        * A `catch` expression for handling Java exceptions.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:947]: expression -> "block" (they are definitely not expressions)

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 6]
```diff
@@ -910,14 +909,54 @@ object ParsedAst {
...
+        *
+        * @param rules the catch rules.
+        */
+      case class Catch(rules: Seq[ParsedAst.CatchRule]) extends CatchOrHandler
+
+      /**
+        * A `with` expression for handling Flix effects.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:954]: expression -> "block" (they are definitely not expressions)

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala [hunk 7]
```diff
@@ -838,17 +838,37 @@ class Parser(val source: Source) extends org.parboiled2.Parser {
...
+      def CatchBody: Rule1[ParsedAst.Expression.CatchOrHandler] = rule {
+        keyword("catch") ~ optWS ~ "{" ~ optWS ~ oneOrMore(CatchRule).separatedBy(CaseSeparator) ~ optWS ~ "}" ~> ParsedAst.Expression.CatchOrHandler.Catch
+      }
+
+      def HandlerRule: Rule1[ParsedAst.HandlerRule] = rule {
+        keyword("def") ~ WS ~ Names.QualifiedEffect ~ FormalParamList ~ optWS ~ atomic("=") ~ optWS ~ Expression ~> ParsedAst.HandlerRule
+      }
+
+      def HandlerBody: Rule1[ParsedAst.Expression.CatchOrHandler] = rule {
+        keyword("with") ~ optWS ~ "{" ~ optWS ~ oneOrMore(HandlerRule).separatedBy(CaseSeparator) ~ optWS ~ "}" ~> ParsedAst.Expression.CatchOrHandler.Handler
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Parser.scala:863]: Can an effect have zero operations? I guess not hence why you require at least one handler? But still, why not allow zero handlers? It should be detected later.

**Comment by magnus-madsen** [comment]
@mlutze Added a few random comments.

**Pull request merged**